### PR TITLE
Make call to std::abort() optional. (#327)

### PR DESCRIPTION
**Pull request by dopiera**
Add {Get,Set}TerminateHandler function which mimic std::set_terminate.
If exceptions are not enabled via (GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS),
the handler installed by those is called instead of plain abort. Default
behaviour was not changed.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1287)

### CODE REVIEW

File: google/cloud/CMakeLists.txt [hunk 1]
```diff
@@ -65,6 +65,8 @@ configure_file(internal/build_info.cc.in internal/build_info.cc)
 
 # the client library
 add_library(google_cloud_cpp_common
+            error.h
```
Comment (coryan) [inline google/cloud/CMakeLists.txt]: Please rename these files to `terminate_handler.h` and `terminate_handler.cc`.

File: google/cloud/error.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ * will result in calling the handler installed via this API. By default,
+ * a call to std::abort() is used.
+ */
+
+using TerminateHandler = std::function<void()>;
+
+/**
+ * Install terminate handler.
+ *
+ * @f The handler. It should never return, behaviour is undefined if it does.
```
Comment (coryan) [inline google/cloud/error.h]: I think you mean `@param f`.

File: google/cloud/error.cc [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    return f_;
+  }
+
+  void Set(TerminateHandler&& f) {
+    std::lock_guard<std::mutex> l(m_);
+    f_ = std::move(f);
+  }
+
+ private:
+  TerminateHandler f_;
```
Comment (coryan) [inline google/cloud/error.cc]: nit: I prefer to see the mutex first, in the future we will want to use the `GUARDED_BY` macros to describe what mutexes protect what.

File: google/cloud/error.cc [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+namespace google {
+namespace cloud {
+inline namespace GOOGLE_CLOUD_CPP_NS {
+namespace {
+
+class TerminateFunction {
+ public:
+  TerminateFunction(TerminateHandler&& f) {
+    std::lock_guard<std::mutex> l(m_);
```
Comment (coryan) [inline google/cloud/error.cc]: nit: you do not need to grab the mutex in the constructor, but it works... meh?  No action needed.

File: google/cloud/error.h [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+ * never return.
+ */
+using TerminateHandler = std::function<void(const char* msg)>;
+
+/**
+ * Install terminate handler.
+ *
+ * @f The handler. It should never return, behaviour is undefined if it does.
+ */
+void SetTerminateHandler(TerminateHandler&& f);
```
Comment (coryan) [inline google/cloud/error.h]: Does this have to be a ref-ref?  `std::function` have value semantics, so a plain by-value would work.  And the application developer can save the copies by using `std::move`.  Thoughts?

Comment (coryan) [inline google/cloud/error.h]: Should `SetTerminateHandler()` return the previous handler atomically?  Just a thought...

**Comment by coryan** [comment]
Typically it is easier to just keep adding commits to the feature branch. At the end we squash all the commits into a single commit to `master`.

From time to time you will need to rebase your commits, for example because something else got committed to `master` that conflicts, or maybe the builds are having trouble and you need to pick up the changes that fix them.

File: google/cloud/CMakeLists.txt (hunk 1 above)
Comment (dopiera) [reply google/cloud/CMakeLists.txt]: Done

File: google/cloud/error.cc (hunk 4 above)
Comment (dopiera) [reply google/cloud/error.cc]: True, looks like I didn't code for too long...

File: google/cloud/error.cc (hunk 3 above)
Comment (dopiera) [reply google/cloud/error.cc]: Done

File: google/cloud/error.h (hunk 2 above)
Comment (dopiera) [reply google/cloud/error.h]: I did, fixed.

File: google/cloud/error.h (hunk 5 above)
Comment (dopiera) [reply google/cloud/error.h]: It certainly doesn't fixed here and in the holder.

Comment (dopiera) [reply google/cloud/error.h]: Good idea, added and tested.

File: google/cloud/terminate_handler.h [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+ * Get the currently installed handler.
+ *
+ * @return The currently installed handler.
+ */
+TerminateHandler GetTerminateHandler();
+
+/**
+ * Invoke the currently installed handler.
+ *
+ * @msg Details about the error.
```
Comment (coryan) [inline google/cloud/terminate_handler.h]: `s/@msg/@param msg/`

File: google/cloud/terminate_handler_test.cc [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+TEST(TerminateHandler, UnsetTerminates) {
+  GetTerminateHandler();
+  EXPECT_DEATH_IF_SUPPORTED(Terminate("Test"),
+                            "Aborting because exceptions are disabled: Test");
+}
+
+TEST(TerminateHandler, SettingGettingWorks) {
+  SetTerminateHandler(&CustomHandler);
+  TerminateHandler set_handler = GetTerminateHandler();
+  ASSERT_EQ(CustomHandler, *set_handler.target<void (*)(const char*)>());
```
Comment (coryan) [inline google/cloud/terminate_handler_test.cc]: I think using `ASSERT_*` will require types that can be printed, you may have to do something like:

```C++
ASSERT_TRUE(CustomHandler == ...) << "The handler objects should be equal";
```

File: google/cloud/internal/throw_delegate.h [hunk 8]
```diff
@@ -27,8 +27,9 @@ namespace internal {
  * @name Delete exception raising to hidden functions.
  *
  * The following functions raise the corresponding exception, unless the user
- * has disabled exception handling, in which case they print the error message
- * to std::cerr and call std::abort().
+ * has disabled exception handling, in which case they call the function set by
+ * google::cloud::SetTerminateHandler() (by default print the error message
```
Comment (coryan) [inline google/cloud/internal/throw_delegate.h]: Consider:

```C++
/*
...
 * `google::cloud::SetTerminateHandler()`. The default handler prints the error
 * message to `std::cerr` and calls `std::abort()`.
*/
```

Note that Doxygen uses markdown formatting so things like code look better with the backticks.

File: google/cloud/internal/throw_delegate.h [hunk 9]
```diff
@@ -27,8 +27,9 @@ namespace internal {
  * @name Delete exception raising to hidden functions.
```
Comment (coryan) [inline google/cloud/internal/throw_delegate.h]: I probably wrote that comment and it makes no sense to me.  Can you change it?

```C++
/**
 * @name Delegate exception raising to out of line functions.
...
```

File: google/cloud/terminate_handler.h (hunk 6 above)
Comment (dopiera) [reply google/cloud/terminate_handler.h]: Ooops, fixed.

File: google/cloud/internal/throw_delegate.h (hunk 9 above)
Comment (dopiera) [reply google/cloud/internal/throw_delegate.h]: Done.

File: google/cloud/internal/throw_delegate.h (hunk 8 above)
Comment (dopiera) [reply google/cloud/internal/throw_delegate.h]: Done

File: google/cloud/terminate_handler_test.cc (hunk 7 above)
Comment (dopiera) [reply google/cloud/terminate_handler_test.cc]: Fascinating, functions are printable in gcc, no idea why MSVC doesn't like them, but probably not worth the investigation.

File: google/cloud/terminate_handler.h [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_TERMINATE_HANDLER_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_TERMINATE_HANDLER_H_
+
+#include <functional>
+
+namespace google {
+namespace cloud {
+inline namespace GOOGLE_CLOUD_CPP_NS {
```
Comment (coryan) [inline google/cloud/terminate_handler.h]: Ugh, I should have noticed this before: `GOOGLE_CLOUD_CPP_NS` is a macro defined in `google/cloud/version.h`, you need to include the file or this header means different things depending on whether that `version.h` was already included or not.

**Pull request merged**